gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/rules/ExecutionEngineArbiter.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.pivotal.gemfirexd.internal.impl.sql.rules;
+
+import java.util.LinkedList;
+
+import com.pivotal.gemfirexd.internal.engine.GemFireXDQueryObserver;
+import com.pivotal.gemfirexd.internal.engine.GemFireXDQueryObserverHolder;
+import com.pivotal.gemfirexd.internal.engine.GfxdConstants;
+import com.pivotal.gemfirexd.internal.engine.distributed.metadata.DMLQueryInfo;
+import com.pivotal.gemfirexd.internal.iapi.services.property.PropertyUtil;
+import com.pivotal.gemfirexd.internal.impl.sql.rules.ExecutionEngineRule.ExecutionEngine;
+
+public class ExecutionEngineArbiter {
+  LinkedList<ExecutionEngineRule> executionEngineRules = new LinkedList<>();
+
+
+  public ExecutionEngineArbiter() {
+    //Rules That needs to be applied regardless of the GFXD_ROUTE_SELECTED_STORE_QUERIES_TO_SPARK flag
+    // mostly for spark queries.
+    executionEngineRules.add(new ColumnTableExecutionEngineRule());
+
+    //Rules that applies  on the store Queries
+    if (Boolean.parseBoolean(
+        PropertyUtil.getSystemProperty(
+            GfxdConstants.GFXD_ROUTE_SELECTED_STORE_QUERIES_TO_SPARK , "true"))) {
+      executionEngineRules.add(new ReplicatedTableExecutionEngineRule());
```

Comment:
On similar lines of creating less temporary objects. Try and make all the rules as static single objects and pass the queryInfo as arguments as you are doing.